zips/zip-0032.rst

```diff
@@ -513,43 +525,119 @@ valid diversifiers.
 
-- $\mathsf{Arbitrary.MKGDomain} = \texttt{“ZcashArbitraryKD”}$
-- $\mathsf{Arbitrary.CKDDomain} = \mathtt{0xAB}$
+- $\mathsf{Registered.MKGDomain} = \texttt{“ZIPRegistered\_KD”}$
+- $\mathsf{Registered.CKDDomain} = \mathtt{0xAC}$
+
+Registered subtree root key generation
+--------------------------------------
+
+Let $S$ be a seed byte sequence of a chosen length, which MUST be at least 32 and at most 252 bytes.
 
-Arbitrary master key generation
+The registered subtree root is obtained as:
+
+$\mathsf{RegKD}(\mathsf{ContextString}, \mathsf{S}, \mathsf{ZipNumber})$ $\rightarrow (\mathsf{sk}_{subtree}, \mathsf{c}_{subtree})$ :
+
+- Let $(\mathsf{sk}_{m}, \mathsf{c}_{m}) = \mathsf{MKGh}^\mathsf{Registered}([\mathsf{length}(\mathsf{ContextString})]\,||\,\mathsf{ContextString}\,||\,[\mathsf{length}(S)]\,||\,S)$.
+- Return $\mathsf{CKDh}^\mathsf{Registered}((\mathsf{sk}_{m}, \mathsf{c}_{m}), \mathsf{ZipNumber} + 2^{31}, [\,], 0)$.
+
+Note: The intermediate key $(\mathsf{sk}_{m}, \mathsf{c}_{m})$ would grant the ability to derive the
+subtree root for application protocols defined in *any* ZIP using the same $\mathsf{ContextString}$
+and seed. This is a potentially dangerous scope of grant since it cannot be known what future
+protocols will use this mechanism; therefore, $(\mathsf{sk}_{m}, \mathsf{c}_{m})$ SHOULD NOT be
+used or stored directly without careful consideration of security consequences.
+
+Registered child key derivation
 -------------------------------
 
+As well as the integer child index $i$, the child key derivation function defined here supports a
+byte sequence $\mathsf{tag}$; each pair $(i, \mathsf{tag})$ produces a different child key. If an
+explicit tag is not required, $\mathsf{tag}$ is set to the empty byte sequence.
+
+Note: The tag MUST use an unambiguous encoding within the given context. It is RECOMMENDED to use
+a prefix-free encoding, which may require the use of length fields if multiple fields need to be
+encoded.
+
+$\mathsf{CKDreg}((\mathsf{sk}_{par}, \mathsf{c}_{par}), i, \mathsf{tag})$ $\rightarrow (\mathsf{sk}_{child}, \mathsf{c}_{child})$ :
+
+- Return $\mathsf{CKDh}^\mathsf{Arbitrary}((\mathsf{sk}_{par}, \mathsf{c}_{par}), i, \mathsf{tag}, 0)$.
+
+Full-width child cryptovalue derivation
+---------------------------------------
+
+If the application protocol requires a 64-byte cryptovalue (for example, to avoid an entropy
+bottleneck in its subsequent operations), then a similar process to `Registered child key derivation`_
+above can be used to obtain such a cryptovalue at a given child index; again with optional tag.
+The same considerations about encoding of the $\mathsf{tag}$ apply as above.
+
+$\mathsf{CKDfw}((\mathsf{sk}_{par}, \mathsf{c}_{par}), i, \mathsf{tag})$ $\rightarrow \mathsf{K}_{child}$ :
+
+- Let $(I_L, I_R) = \mathsf{CKDh}^\mathsf{Registered}((\mathsf{sk}_{par}, \mathsf{c}_{par}), i, \mathsf{tag}, 1)$.
+- Return $I_L\,||\,I_R$.
```

Comment:
I can see why it's okay to allow derivation of children below this path element `i`, but I'd prefer text making it explicit:
```suggestion
- Return $I_L\,||\,I_R$.

For the avoidance of doubt, the output of $\mathsf{CKDfw}$ MUST NOT be provided
as input to any key derivation function defined in ZIP 32. In particular, it
MUST NOT be reinterpreted as an input to $\mathsf{CKDreg}$.
```
or similar (basically enforcing that the way you derive children along ZIP 32 paths must be via `CKDreg`).